### ISSUE TITLE
no-unused-variable error with object destructuring

### ISSUE THREAD

**Issue by cartant** [palantir/tslint#2876]
### Bug Report

- __TSLint version__: 5.4.2
- __TypeScript version__: 2.3.4
- __Running TSLint via__: CLI

#### TypeScript code being linted

To reproduce the problem, two files are required:

```ts
// main.ts
import { SomeClass } from "./some";
const person = { name: "alice" };
const { name } = person;
```

```ts
// some.ts
export class SomeClass {};
```

with `tslint.json` configuration:

```json
{
    "defaultSeverity": "error",
    "rules": { "no-unused-variable": { "severity": "error" } }
}
```

#### Actual behavior

An internal error within the rule occurs:

```
Warning: Cannot read property 'flags' of undefined
```

#### Expected behavior

TSLint should execute without internal errors and should report that `SomeClass` and `name` are both unused.

#### Additional information

The import is required to effect the internal error. If the import is omitted, no internal error occurs.

With the import, the rule follows a code path into the [`getImplicitType`](https://github.com/palantir/tslint/blob/5.4.2/src/rules/noUnusedVariableRule.ts#L277) function and the implementations of `utils.isPropertyDeclaration` and `utils.isVariableDeclaration` allow for the `{ name } = person` node to be passed to `checker.getTypeAtLocation` - from within which the error is thrown.

**Comment by victornoel** [comment]
@cartant do you think this is the same as #3001?

**Comment by cartant** [comment]
@victornoel Maybe. That issue doesn't seem to include enough information for the reported problem to be reproduced, so who knows?

**Comment by victornoel** [comment]
@adidahiya any plan to fix this issue? it is really problematic…

**Comment by adidahiya** [comment]
Is this still an issue with TSLint 5.5.0?

**Comment by cartant** [comment]
@adidahiya I've revisted the harness I created and the error is still effected with:

- __TSLint version__: 5.5.0
- __TypeScript version__: 2.4.1
- __Running TSLint via__: CLI

---

    TypeError: Cannot read property 'flags' of undefined
        at Object.getCheckFlags (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:10137:22)
        at getTypeOfSymbol (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:30957:20)
        at getTypeOfNode (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:47329:24)
        at Object.getTypeAtLocation (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:26823:31)
        at getImplicitType (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\tslint\lib\rules\noUnusedVariableRule.js:269:24)
        at cb (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\tslint\lib\rules\noUnusedVariableRule.js:259:20)
        at visitEachNode (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:15956:30)
        at Object.forEachChild (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:16143:24)
        at cb (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\tslint\lib\rules\noUnusedVariableRule.js:264:19)
        at visitNode (C:\Users\Nicholas\git\temp\tslint-no-unused-variable\node_modules\typescript\lib\typescript.js:15944:20)

**Comment by victornoel** [comment]
I concur :)

**Comment by SteveALee** [comment]
For information I had same hard to locate warning with "no-use-before-declare": true, in TSL 4.5.1 

It would really help if all TSL errors were prefixed with [TypeScript Lint] or similar.

**Comment by victornoel** [comment]
@SteveALee open another issue for this please…

**Comment by SteveALee** [comment]
@victornoel your mean for the prefix?

**Comment by victornoel** [comment]
@SteveALee yes, I meant that if you have a need for prefixing tslint messages, open another issue for that

**Comment by SteveALee** [comment]
@victornoel please see #3060 :)
